x/upgrade/abci.go

```diff
@@ -32,66 +39,76 @@ func BeginBlocker(k *keeper.Keeper, ctx sdk.Context) {
 		// 1. If there is no scheduled upgrade.
 		// 2. If the plan is not ready.
 		// 3. If the plan is ready and skip upgrade height is set for current height.
-		if !found || !plan.ShouldExecute(ctx) || (plan.ShouldExecute(ctx) && k.IsSkipHeight(ctx.BlockHeight())) {
-			lastAppliedPlan, _ := k.GetLastCompletedUpgrade(ctx)
+		if !found || !plan.ShouldExecute(blockHeight) || (plan.ShouldExecute(blockHeight) && k.IsSkipHeight(blockHeight)) {
+			lastAppliedPlan, _, err := k.GetLastCompletedUpgrade(ctx)
+			if err != nil {
+				return err
+			}
+
 			if lastAppliedPlan != "" && !k.HasHandler(lastAppliedPlan) {
 				var appVersion uint64
 
-				cp := ctx.ConsensusParams()
+				cp := sdkCtx.ConsensusParams()
 				if cp.Version != nil {
 					appVersion = cp.Version.App
 				}
 
-				panic(fmt.Sprintf("Wrong app version %d, upgrade handler is missing for %s upgrade plan", appVersion, lastAppliedPlan))
+				return fmt.Errorf("wrong app version %d, upgrade handler is missing for %s upgrade plan", appVersion, lastAppliedPlan)
 			}
 		}
 	}
 
 	if !found {
-		return
+		return nil
 	}
-	logger := ctx.Logger()
+
+	logger := k.Logger(ctx)
 
 	// To make sure clear upgrade is executed at the same block
-	if plan.ShouldExecute(ctx) {
+	if plan.ShouldExecute(blockHeight) {
 		// If skip upgrade has been set for current height, we clear the upgrade plan
-		if k.IsSkipHeight(ctx.BlockHeight()) {
+		if k.IsSkipHeight(blockHeight) {
 			skipUpgradeMsg := fmt.Sprintf("UPGRADE \"%s\" SKIPPED at %d: %s", plan.Name, plan.Height, plan.Info)
 			logger.Info(skipUpgradeMsg)
 
 			// Clear the upgrade plan at current height
-			k.ClearUpgradePlan(ctx)
-			return
+			return k.ClearUpgradePlan(ctx)
 		}
 
 		// Prepare shutdown if we don't have an upgrade handler for this upgrade name (meaning this software is out of date)
 		if !k.HasHandler(plan.Name) {
 			// Write the upgrade info to disk. The UpgradeStoreLoader uses this info to perform or skip
 			// store migrations.
-			err := k.DumpUpgradeInfoToDisk(ctx.BlockHeight(), plan)
+			err := k.DumpUpgradeInfoToDisk(blockHeight, plan)
 			if err != nil {
-				panic(fmt.Errorf("unable to write upgrade info to filesystem: %s", err.Error()))
+				return fmt.Errorf("unable to write upgrade info to filesystem: %s", err.Error())
 			}
 
 			upgradeMsg := BuildUpgradeNeededMsg(plan)
 			logger.Error(upgradeMsg)
-			panic(upgradeMsg)
+
+			// Returning an error will end up in a panic
+			return fmt.Errorf(upgradeMsg)
 		}
 
 		// We have an upgrade handler for this upgrade name, so apply the upgrade
-		ctx.Logger().Info(fmt.Sprintf("applying upgrade \"%s\" at %s", plan.Name, plan.DueAt()))
-		ctx = ctx.WithBlockGasMeter(storetypes.NewInfiniteGasMeter())
-		k.ApplyUpgrade(ctx, plan)
-		return
+		logger.Info(fmt.Sprintf("applying upgrade \"%s\" at %s", plan.Name, plan.DueAt()))
+		sdkCtx = sdkCtx.WithBlockGasMeter(storetypes.NewInfiniteGasMeter())
+		// TODO: figure out how to pass a block gas meter to the ApplyUpgradeFunction
```

Comment:
Is this done in a follow-up? Or in this PR?

Reply:
we should remove this todo, its not needed